source/proxy_connection.c

```diff
@@ -0,0 +1,316 @@
+
+    int result = AWS_OP_ERR;
+
+    if (aws_byte_buf_init(
+            &base64_input_value,
+            proxy_user_data->allocator,
+            proxy_user_data->username->len + proxy_user_data->password->len + 1)) {
+        goto done;
+    }
+
+    /* First build a buffer with "username:password" in it */
+    struct aws_byte_cursor username_cursor = aws_byte_cursor_from_string(proxy_user_data->username);
+    if (aws_byte_buf_append(&base64_input_value, &username_cursor)) {
+        goto done;
+    }
+
+    struct aws_byte_cursor colon_cursor = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL(":");
+    if (aws_byte_buf_append(&base64_input_value, &colon_cursor)) {
+        goto done;
+    }
+
+    struct aws_byte_cursor password_cursor = aws_byte_cursor_from_string(proxy_user_data->password);
+    if (aws_byte_buf_append(&base64_input_value, &password_cursor)) {
+        goto done;
+    }
+
+    struct aws_byte_cursor base64_source_cursor =
+        aws_byte_cursor_from_array(base64_input_value.buffer, base64_input_value.len);
+
+    /* Figure out how much room we need in our final header value buffer */
+    size_t required_size = 0;
+    if (aws_base64_compute_encoded_len(base64_source_cursor.len, &required_size)) {
+        goto done;
+    }
+
+    required_size += s_proxy_authorization_header_basic_prefix->len + 1;
+    if (aws_byte_buf_init(&header_value, proxy_user_data->allocator, required_size)) {
+        goto done;
+    }
+
+    /* Build the final header value by appending the authorization type and the base64 encoding string together */
+    struct aws_byte_cursor basic_prefix = aws_byte_cursor_from_string(s_proxy_authorization_header_basic_prefix);
+    if (aws_byte_buf_append_dynamic(&header_value, &basic_prefix)) {
+        goto done;
+    }
+
+    if (aws_base64_encode(&base64_source_cursor, &header_value)) {
+        goto done;
+    }
+
+    struct aws_http_header header = {.name = aws_byte_cursor_from_string(s_proxy_authorization_header_name),
+                                     .value = aws_byte_cursor_from_array(header_value.buffer, header_value.len)};
+
+    if (aws_http_message_add_header(request, header)) {
+        goto done;
+    }
+
+    result = AWS_OP_SUCCESS;
+
+done:
+
+    aws_byte_buf_clean_up(&header_value);
+    aws_byte_buf_clean_up(&base64_input_value);
+
+    return result;
+}
+
+AWS_STATIC_STRING_FROM_LITERAL(s_http_scheme, "http");
+
+static int s_rewrite_uri_for_proxy_request(
+    struct aws_http_message *request,
+    struct aws_http_proxy_user_data *proxy_user_data) {
+    int result = AWS_OP_ERR;
+
+    struct aws_uri target_uri;
+    AWS_ZERO_STRUCT(target_uri);
+
+    struct aws_byte_cursor path_cursor;
+    AWS_ZERO_STRUCT(path_cursor);
+
+    if (aws_http_message_get_request_path(request, &path_cursor)) {
+        goto done;
+    }
+
+    /* Pull out the original path/query */
+    struct aws_uri uri;
+    if (aws_uri_init_parse(&uri, proxy_user_data->allocator, &path_cursor)) {
```

Comment:
An OPTIONS request can have "*" for the path. What happens then?

Reply:
Shouldn't cause any problems.  I did a few experiments using curl through a proxy with and without '*' and there was no difference in responses given to the OPTIONS request.

Reply:
But "*" isn't a valid URI, so this parse() call will fail, so the OPTIONS request will fail to even send